kpeerset/peerheap/heap.go

```diff
@@ -0,0 +1,110 @@
+package peerheap
+
+import (
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// Comparator is the type of a function that compares two peer Heap items to determine the ordering between them.
+// It returns true if i1 is "less" than i2 and false otherwise.
+type Comparator func(i1 Item, i2 Item) bool
+
+// Item is one "item" in the Heap.
+// It contains the Id of the peer, an arbitrary value associated with the peer
+// and the index of the "item" in the Heap.
+type Item struct {
+	Peer  peer.ID
+	Value interface{}
+	Index int
```

Comment:
this is a very generic 'item' interface, but it looks like there's only one concrete use of the heap right now - which uses an integer metric as the ranking of known peer IDs. do we have a second use case that would have a different type for Value that warrants the abstraction?

Reply:
@willscott As of _now_, we do not have a second use case. However, I do see us using the heap more & more as we test with different metrics for peer selection and would like to keep this as is unless you have strong objections. Limiting users of the Heap to `big.Int` seems like a strong ask.